### ISSUE TITLE
Publishing workflows

### ISSUE THREAD

**Issue by guybedford** [jspm/jspm-cli#249]
How to write a jspm package using jspm and then publish to npm?

**Comment by guybedford** [comment]
For now this is not a priority, but an Endpoint publish hook is certainly something that can be integrated in future.

**Comment by guybedford** [comment]
Note that it is worth supporting the `private` field in npm as well, see https://github.com/jspm/jspm-cli/issues/588.

**Comment by interactionist** [comment]
I hope ```jspm``` will stay a top-notch bundler without the capability to publish to a public registry.

**Comment by guybedford** [comment]
To summarize what is needed to make this work:

* `jspm publish` command.
* Reads registry from package.json.
* Calls a "publish" hook for the registry itself.

We can then add a `publish` hook support to both jspm-github and jspm-npm.

Marking as contributions welcome.

**Comment by lookfirst** [comment]
Can you summarize what the `publish` hook would do?

**Comment by guybedford** [comment]
Exactly the same thing as npm publish, but it would call out to the registry implementation to do the actual publishing.

Example:

```json
{
  "name": "username/project",
  "version": "1.2.3",
  "registry": "github"
}
```

```
  jspm publish
```

Might tag a `v1.2.3` and pushes the project to GitHub (assuming there is a local git repo set up).

And similarly for `registry: 'npm'`.

**Comment by peteruithoven** [comment]
Isn't it strange to prefix the `name`  with the github user/company? Isn't that strange. Wouldn't that give issues when publishing to npm? 
I don't see others do this? For example: 
- https://github.com/karma-runner/karma/blob/master/package.json
- https://github.com/strongloop/express/blob/master/package.json
- https://github.com/substack/node-browserify/blob/master/package.json

**Comment by guybedford** [comment]
@peteruithoven see https://github.com/jspm/jspm-cli/issues/249#issuecomment-87449797. The "full name" is `[registry]:[name]@[version]` where those are each separate package.json properties, making it backwards compatible with the npm convention.

**Comment by peteruithoven** [comment]
I might be confused but what I understand is that you require the following "full name": 
```
[registry]:[user name][package name]@[version]
```
For example:
```
github:guybedford/markdown-component@1.0.0
```
I don't get where JSPM is getting the username from. I don't believe it's common to include the username in package.json's `name` property (see my example packages earlier).

**Comment by guybedford** [comment]
@peteruithoven on jspm I usually refer to the _package full name_ as `registry:package@version` and the _package name_ as `package`. That is, the package name is `guybedford/markdown-component` for a package on GitHub. The name and version as written in the package.json is not handled by jspm though unless using publishing or linking workflows, and publishing workflows don't exist yet.

Just let me know if you have any further questions at all.

**Comment by peteruithoven** [comment]
I was mostly thinking about the linking workflow, that's where I got confused. But this topic seemed like a better place to ask this question, because it might also be relevant for the future publishing workflow.

**Comment by guybedford** [comment]
Closing - these future features are better to come back to when ready than leave hanging...
